--- /var/www/ilias51/Customizing/global/plugins/Modules/Cloud/CloudHook/OwnCloud/classes/class.ownclTreeGUI.php 2017-06-29 15:50:23.199348929 +0200
+++ /root/sciebo/OwnCloud-oauth2-1d49bebf50d908f5b054436a86ea5515816cddac/classes/class.ownclTreeGUI.php        2017-06-29 10:44:20.000000000 +0200
@@ -69,9 +69,7 @@
         */
        function getNodeContent($node)
        {
-               $config = new ownclConfig();
-       
-               $node->getName() ? $name = $node->getName() : $name = $config->getServiceTitle();
+               $node->getName() ? $name = $node->getName() : $name = 'OwnCloud';
                return $name;
        }