diff --git a/src/Http/CurlClient.php b/src/Http/CurlClient.php
===================================================================
--- a/src/Http/CurlClient.php	(date 1602668820027)
+++ b/src/Http/CurlClient.php	(date 1602668820027)
@@ -2,6 +2,8 @@

 namespace ceLTIc\LTI\Http;

+use ILIAS\DI\Container;
+
 /**
  * Class to implement the HTTP message interface using the Curl library
  *
@@ -21,6 +23,8 @@
      */
     public function send(HttpMessage $message)
     {
+        global $DIC; /** @var Container $DIC */
+
         $ch = curl_init();
         curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 30);
         curl_setopt($ch, CURLOPT_URL, $message->getUrl());
@@ -41,7 +45,16 @@
         curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
         curl_setopt($ch, CURLINFO_HEADER_OUT, true);
         curl_setopt($ch, CURLOPT_HEADER, true);
+        if(
+            $DIC->clientIni()->variableExists('system', 'DEVMODE') &&
+            (int)$DIC->clientIni()->readVariable('system', 'DEVMODE') === 1
+        ) {
+            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
+            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
+        }
         $chResp = curl_exec($ch);
+        \ilLoggerFactory::getLogger('lti')->dump(curl_getinfo($ch), \ilLogLevel::DEBUG);
+        \ilLoggerFactory::getLogger('lti')->dump(curl_error($ch), \ilLogLevel::DEBUG);
         $message->requestHeaders = trim(str_replace("\r\n", "\n", curl_getinfo($ch, CURLINFO_HEADER_OUT)));
         $chResp = str_replace("\r\n", "\n", $chResp);
         $chRespSplit = explode("\n\n", $chResp, 2);
@@ -63,5 +76,4 @@

         return $message->ok;
     }
-
 }
