diff --git a/library/HTMLPurifier/AttrDef/URI/Host.php b/library/HTMLPurifier/AttrDef/URI/Host.php
index e7df800..bfe48f2 100644
--- a/library/HTMLPurifier/AttrDef/URI/Host.php
+++ b/library/HTMLPurifier/AttrDef/URI/Host.php
@@ -93,7 +93,14 @@ class HTMLPurifier_AttrDef_URI_Host extends HTMLPurifier_AttrDef
         // punycoding them. (This is the most portable thing to do,
         // since otherwise we have to assume browsers support
 
-        if ($config->get('Core.EnableIDNA')) {
+        // mjansen-patch: begin
+        if (function_exists('idn_to_ascii')) {
+            return idn_to_ascii($string);
+            // If we have Net_IDNA2 support, we can support IRIs by
+            // punycoding them. (This is the most portable thing to do,
+            // since otherwise we have to assume browsers support
+        } elseif ($config->get('Core.EnableIDNA')) {
+        // mjansen-patch: end
             $idna = new Net_IDNA2(array('encoding' => 'utf8', 'overlong' => false, 'strict' => true));
             // we need to encode each period separately
             $parts = explode('.', $string);
